docs/source/optimize.rst

```diff
@@ -1,9 +1,182 @@
+    word list has 2 occurrences of orange, out of 7 words
+
+.. image:: images/optimize_dask2.png
+   :width: 30 %
+   :alt: After culling
+
+Looking at the task graph above, there are multiple accesses to constants such
+as ``'val1'`` or ``'val2'`` in the dask. These can be inlined into the tasks to
+improve efficiency using the ``inline`` function.
+
+.. code-block:: python
+
+    >>> from dask.optimize import inline
+    >>> dsk2 = inline(dsk1)
+    >>> results = get(dsk2, ['print1', 'print2'])
+    word list has 2 occurrences of apple, out of 7 words
+    word list has 2 occurrences of orange, out of 7 words
+
+.. image:: images/optimize_dask3.png
+   :width: 20 %
+   :alt: After inlining
+
+Now we have two sets of *almost* linear task chains. The only link between them
+is the word counting function. For cheap operations like this, the
+serialization cost may be larger than the actual computation, so it may be
+faster to do the computation more than once, rather than passing the results to
+all nodes. To perform this function inlining, the ``inline_functions`` function
+can be used.
+
+.. code-block:: python
+
+    >>> from dask.optimize import inline_functions
+    >>> dsk3 = inline_functions(dsk2)
+    >>> results = get(dsk3, ['print1', 'print2'])
+    word list has 2 occurrences of apple, out of 7 words
+    word list has 2 occurrences of orange, out of 7 words
+
+.. image:: images/optimize_dask4.png
+   :width: 20 %
+   :alt: After inlining functions
+
+Now we have a set of purely linear tasks. We'd like to have the scheduler run
+all of these on the same worker to reduce data serialization between workers.
+One option is just to merge these linear chains into one big task using the
+``fuse`` function.
+
+.. code-block:: python
+
+    >>> from dask.optimize import fuse
+    >>> dsk4 = fuse(dsk3)
+    >>> results = get(dsk4, ['print1', 'print2'])
+    word list has 2 occurrences of apple, out of 7 words
+    word list has 2 occurrences of orange, out of 7 words
+
+.. image:: images/optimize_dask5.png
+   :width: 20 %
+   :alt: After fusing
+
+
+Putting it all together:
+
+.. code-block:: python
+
+    >>> def optimize_and_get(dsk, keys):
+            dsk1 = cull(dsk, keys)
+            dsk2 = inline(dsk1)
+            dsk3 = inline_functions(dsk2, [len, str.split])
+            dsk4 = fuse(dsk2)
+            return get(dsk4, keys)
+    >>> optimize_and_get(dsk, ['print1', 'print2'])
+    word list has 2 occurrences of apple, out of 7 words
+    word list has 2 occurrences of orange, out of 7 words
+
+
+In summary, the above operations:
+
+1. Removed tasks unncessary for the desired output using ``cull``
+2. Inlined constants using ``inline``
+3. Inlined cheap computations using ``inline_functions``, improving parallelism
+4. Fused linear tasks together to ensure they run on the same worker, using ``fuse``
+
+These are just a few of the optimizations provided in ``dask.optimize``, for
+more information see the api below.
+
```

Comment:
Might want to add that the dask collections (array, bag, dataframe) automatically apply a well chosen set of optimizations.  Ordinary users of dask collections don't need to manage optimizations.